ykcompile/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+#![feature(proc_macro_hygiene)]
+#![feature(test)]
+
+#[macro_use]
+extern crate dynasm;
+extern crate dynasmrt;
+extern crate test;
+
+use std::collections::HashMap;
+use std::mem;
+
+use yktrace::tir::{TirTrace, TirOp, Statement, Rvalue, Operand, Constant, ConstantInt,
+UnsignedInt};
+
+use dynasmrt::DynasmApi;
+
+/// A compiled SIRTrace.
+pub struct CompiledTrace {
+    /// A compiled trace.
+    mc: dynasmrt::ExecutableBuffer
+}
+
+impl CompiledTrace {
+    pub fn execute(&self) -> u64 {
+        // For now a compiled trace always returns whatever has been left in register RAX. We also
+        // assume for now that this will be a u64.
+        let func: fn() -> u64 = unsafe {
+            mem::transmute(self.mc.ptr(dynasmrt::AssemblyOffset(0)))
+        };
+        func()
+    }
+}
+
+/// The TraceCompiler takes a SIRTrace and compiles it to machine code. Returns a CompiledTrace.
+pub struct TraceCompiler {
+    asm: dynasmrt::x64::Assembler,
```

Comment:
Should we build our API so that adding other architectures later will be easier?

For example, should we parametrise the type of the `asm` field?

Reply:
We can do it now or we can do it later. I don't think we'll care about other architectures for a while.

Reply:
Yeah. I'm just wondering if we can avoid the tedium of having someone go over a (possibly large) existing code-base later, abstracting all of the x86-specifics with an interface.

Reply:
Since this PR has already been going for a while I suggest we get it merged and leave this for a later PR.